app/packs/src/apps/mydb/elements/details/samples/SampleDetails.js

```diff
@@ -739,59 +748,121 @@ export default class SampleDetails extends React.Component {
     )
   }
 
-  addManualCas(e) {
-    DetailActions.updateMoleculeCas(this.props.sample, e.value);
+  // addManualCas(e) {
+  //   DetailActions.updateMoleculeCas(this.props.sample, e.value);
+  // }
+
+  addHypenToCas(cas, boolean) {
+    const { sample } = this.state;
+    const filterCas = cas.replace(/-/g, '');
+    const firstHypenIndex = filterCas.length - 3;
+    const secondHypenIndex = filterCas.length - 1;
+    console.log(filterCas);
+    const firstPart = filterCas.substring(0, firstHypenIndex);
+    const secondPart = filterCas.substring(firstHypenIndex, secondHypenIndex);
+    const thirdPart = filterCas.substring(secondHypenIndex, filterCas.length);
+    const format = `${firstPart}-${secondPart}-${thirdPart}`;
+    if (boolean) {
+      sample.xref = { ...sample.xref, cas: format };
+      this.setState({ sample });
+    }
+    console.log(format);
+    const checkFormat = format.match(/([0-9]{2,7})-([0-9]{2})-[0-9]/);
+    console.log(checkFormat);
+    return checkFormat;
+  }
+
+  checkCasValidation(cas, boolean) {
+    this.setState({ validateCas: true });
+    const filterCas = cas.replace(/-/g, '');
+    if (cas && cas !== '') {
+      let final = 0;
+      const isnum = /^[0-9-]+$/.test(cas) && !(/^[-]+$/.test(cas));
+      let match;
+      if (filterCas.length >= 5 && isnum) {
+        const checkFormat = cas.match(/([0-9]{2,7})-([0-9]{2})-[0-9]/);
+        match = checkFormat || this.addHypenToCas(cas, boolean);
+        const digits = (match[1] + match[2]).split('').reverse();
+        let sum = 0;
+        for (let i = 0; i < digits.length; i += 1) {
+          sum += (i + 1) * parseInt(digits[i], 10);
+        }
+        final = sum % 10;
```

Comment:
It would be much easier to understand, if you could extract the content from line 783 to 790 in a function with a meaningful name. 
The extracted method can then also be easier tested.

Reply:
done. if you want to dive more into cas validation check this
https://metamolecular.com/blog/2013/10/11/validating-cas-numbers-in-javascript/